src/main.rs

```diff
@@ -20,49 +22,62 @@ fn configure_and_try_main() -> Result<()> {
 }
 
 fn try_main(config: Config) -> Result<()> {
-    let pid_file = config.pid_file.to_owned();
-    if let Err(err) = lock(&pid_file) {
-        return Err(anyhow!(
-            "Failed to create lock file {:?}: {} (tip: use --pid-file to \
-            change the location of the lock file) ",
-            &pid_file,
-            err
-        ));
-    }
+    let mut lock_file = lock(&config)?;
+    write_pid_file(&config.pid_file)?;
 
-    let process_res = process(config);
-    if let Err(e) = unlock(&pid_file) {
-        eprint!("Failed to remove pid file {:?}: {}", pid_file, e);
-    }
+    let process_res = process(&config);
+
+    lock_file.unlock().with_context(|| "Could not remove lockfile")?;
+    remove_pid_file(&config.pid_file)?;
 
     process_res
 }
 
-pub fn lock(pid_file: &Path) -> Result<()> {
-    if pid_file.is_file() {
-        return Err(anyhow!("Lock file {:?} exists, aborting", &pid_file));
+fn lock(config: &Config) -> Result<LockFile> {
+    if !config.state_dir.exists() {
+        debug!("State directory '{:?}' does not exist yet, will try to create it.", config.state_dir);
+        std::fs::create_dir_all(&config.state_dir)
+            .with_context(|| format!("Cannot create state directory: {:?}", config.state_dir))?;
+    }
+
+    let lock_file_path = config.lock_file();
+    let mut lock_file = LockFile::open(&lock_file_path)
+        .with_context(|| format!("Cannot open lockfile: {:?}", lock_file_path))?;
+    
+    if !lock_file.try_lock()
+        .with_context(|| format!("Cannot lock using lockfile: {:?}", lock_file_path))?
+    {
+        Err(anyhow!(format!("another krill-sync process holds the lock at {:?}", lock_file_path)))
+    } else {
+        Ok(lock_file)
     }
+}
+
+fn write_pid_file(pid_file: &Path) -> Result<()> {
     write_buf(
-        &pid_file,
+        pid_file,
         &format!("{}\n", std::process::id()).as_bytes().to_vec(),
-    )?;
-
-    // Ensure the lock file is removed even if we are killed by SIGINT or SIGTERM
-    let unlock_pid_file = pid_file.to_owned();
+    ).with_context(|| format!("Cannot create pid file: {:?}, (tip use --pid_file to change the location).", pid_file))?;
+    
+    // Ensure the pid file is removed even if we are killed by SIGINT or SIGTERM
```

Comment:
Shouldn't we also ensure we unlock the lock file, so this ctrl-c functionality is no longer just about the PID file? (or as stated above get rid of the PID file (but keep the CTRL-C check) or lock the PID file and get rid of the separate lock file) ?

Reply:
No actually, it will unlock when it goes out of scope. So doing that now..

The actual file remains. But, when acquiring the lock the lib will check whether another process holds a file descriptor for the lock file, so it will be unlocked even if ctrl-c was used.

Reply:
Ah cool!